src/theme/BlogPostItem/index.js

```diff
@@ -0,0 +1,38 @@
+import React, { useEffect, useState } from 'react';
+import { DiscussionEmbed } from 'disqus-react';
+import BlogPostItem from '@theme-original/BlogPostItem';
+import { useBlogPost } from '@docusaurus/plugin-content-blog/client';
+import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
+import { useColorMode } from '@docusaurus/theme-common';
+
+export default function BlogPostItemWrapper(props) {
+  const { metadata } = useBlogPost();
+  const { frontMatter, permalink, title } = metadata;
+  const { comments = true } = frontMatter;
+  const { siteConfig, i18n } = useDocusaurusContext();
+  const shortname = siteConfig.customFields.shortname;
+  const { colorMode } = useColorMode();
+  const [key, setKey] = useState(0);
+
+  useEffect(() => {
+    setKey((prevKey) => prevKey + 1);
+  }, [colorMode]);
```

Comment:
We can do it less obscure:

```javascript
  const [disqusColorMode, setDisqusColorMode] = useState(colorMode);

  useEffect(() => {
    setDisqusColorMode((_) => colorMode);
  }, [colorMode]);
``` 

Then we can pass `disqusColorMode` to `DiscussionEmbed` instead of `key` and it will re-render whenever `colorMode` changes.

Reply:
It would be nice to put a comment telling why we need to do this trick.